### PR TITLE
Cmn 1254 add udt

### PR DESCRIPTION
**Pull request by if-loop69420**
## Summary
Add support for parsing user defined types through `CREATE TYPE`-statements

## What's been done?
- Add Tokens
- Add SyntaxKinds
- Add support element_spec
- Add parsing for everything
- Write massive amounts of tests
--- 
Have fun reviewing this massive change. This was so much work.

### CODE REVIEW

File: src/grammar/element_spec.rs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    }
+    p.finish();
+}
+
+fn parse_subprogram_spec(p: &mut Parser) {
+    p.expect_one_of(&[T![member], T![static]]);
+    match p.current() {
+        T![procedure] => parse_procedure_spec(p),
+        T![function] => parse_function_spec(p),
+        _ => (),
```
Comment (KieranKaelin) [inline src/grammar/element_spec.rs:34]: Should there be an error here?

**Review by KieranKaelin — CHANGES_REQUESTED** [review]

File: src/grammar/udt.rs [hunk 2]
```diff
@@ -0,0 +1,1052 @@
+use crate::{
+    grammar::{element_spec::parse_element_spec, parse_datatype},
+    safe_loop, Parser,
+};
+use source_gen::{lexer::TokenKind, syntax::SyntaxKind, T};
+
+use super::{call_spec::opt_call_spec, parse_block, parse_expr, parse_ident, parse_param_list};
+
+#[allow(unused)]
```
[the remark this answers is not in the document]
Comment (if-loop69420) [reply src/grammar/udt.rs:9]: Don't really have anywhere to put this right now sadly

[the remark this answers is not in the document]
Comment (if-loop69420) [reply src/grammar/udt.rs:9]: Other than this issue we can merge if we want

File: src/grammar/udt.rs [hunk 3]
```diff
@@ -0,0 +1,1052 @@
...
+}
+
+fn parse_func_decl_in_type(p: &mut Parser) {
+    p.start(SyntaxKind::FuncDeclInType);
+    p.expect(T![function]);
+    parse_ident(p, 1..2);
+    parse_param_list(p);
+    p.expect(T![return]);
+    parse_datatype(p);
+    // Parse stupid bullshit
```
Comment (KieranKaelin) [inline src/grammar/udt.rs:111]:
```suggestion
```

File: src/grammar/udt.rs [hunk 4]
```diff
@@ -0,0 +1,1052 @@
...
+    , cust_first_name    VARCHAR2(20)
+    , cust_last_name     VARCHAR2(20)
+    , cust_address       CUST_ADDRESS_TYP
+    , phone_numbers      PHONE_LIST_TYP
+    , nls_language       VARCHAR2(3)
+    , nls_territory      VARCHAR2(30)
+    , credit_limit       NUMBER(9,2)
+    , cust_email         VARCHAR2(30)
+    , cust_orders        ORDER_LIST_TYP
+    ) ;",
```
Comment (KieranKaelin) [inline src/grammar/udt.rs:479]: Lets reduce this to the relevant parts:
```suggestion
                "CREATE TYPE customer_typ_demo AS OBJECT ( customer_id NUMBER(6)
    , cust_address       CUST_ADDRESS_TYP
    ) ;",
```

File: src/grammar/udt.rs [hunk 5]
```diff
@@ -0,0 +1,1052 @@
...
+        check(
+            parse(
+                "CREATE TYPE cust_address_typ2 AS OBJECT
+       ( street_address     VARCHAR2(40)
+       , postal_code        VARCHAR2(10)
+       , city               VARCHAR2(30)
+       , state_province     VARCHAR2(10)
+       , country_id         CHAR(2)
+       , phone              phone_list_typ_demo
+       );",
```
Comment (KieranKaelin) [inline src/grammar/udt.rs:624]: Whats the difference to the test above?

File: src/grammar/udt.rs [hunk 6]
```diff
@@ -0,0 +1,1052 @@
...
+        );
+    }
+
+    #[test]
+    fn test_object_sub_type() {
+        check(
+            parse(
+                "CREATE TYPE corporate_customer_typ_demo UNDER customer_typ
+    ( account_mgr_id     NUMBER(6)
+    );",
```
Comment (KieranKaelin) [inline src/grammar/udt.rs:721]:
```suggestion
                "CREATE TYPE corporate_customer_typ_demo UNDER customer_typ (account_mgr_id     NUMBER(6));",
```

**Review by KieranKaelin — CHANGES_REQUESTED** [review]

File: src/grammar/udt.rs (hunk 5 above)
Comment (if-loop69420) [reply src/grammar/udt.rs]: true. This is redundant

**Review by KieranKaelin — APPROVED** [review]

**Pull request merged**